### PR TITLE
[PLT-4340] Channel Mute and "/mute" command #7617

### PR DESCRIPTION
**Pull request by chclaus**
#### Summary
Add `/mute` command.

#### Ticket Link
[Github](https://github.com/mattermost/mattermost-server/issues/7617)

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [ ] All new/modified APIs include changes to the drivers
- [ ] Has enterprise changes (please link)
- [ ] Has UI changes
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json) updates
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

#### Further notes:
I've marked this PR as WIP since the webapp-implementation isn't feature complete: See the [webapp PR](https://github.com/mattermost/mattermost-webapp/pull/196)

### CODE REVIEW

**Comment by jasonblais** [comment]
@chclaus Are the server-side changes ready for review though? We can review the server and webapp PRs separately.

**Comment by jasonblais** [comment]
And a huge thanks for the contribution!

**Comment by chclaus** [comment]
Hej @jasonblais , I enjoyed it 😁 
... and yes, the PR can be reviewed. I've no idea why the tests aren't passing. I found neither a `fail:` nor a `panic` in my console and thought all tests passed well 🤔

**Comment by esethna** [comment]
Thanks @chclaus! Super excited about this! I retriggered the build to see if that solves it. We'll review the server PR separately so I've removed the WIP label

**Comment by chclaus** [comment]
Ok, did some mistakes. Now the test should work ;)

**Comment by chclaus** [comment]
Hej @esethna,

I'm happy that the changes looks good. I18n keys are fixed now. For personal reasons, the webapp PR have to wait a few more days... :(

**Comment by esethna** [comment]
@chclaus please help to resolve the conflicts when you get a chance?

**Comment by chclaus** [comment]
Done ;)

File: i18n/en.json [hunk 1]
```diff
@@ -932,6 +932,34 @@
...
+    "id": "api.command_mute.success_mute",
+    "translation": "You will not receive notifications for {{.Channel}} until channel mute is turned off."
+  },
+  {
+    "id": "api.command_mute.success_unmute",
+    "translation": "{{.Channel}} is no longer muted."
+  },
+  {
+    "id": "api.command_mute.success_mute_direct_msg",
+    "translation": "You will not receive notifications for this direct message channel until channel mute is turned off."
```
Comment (esethna) [inline i18n/en.json]: Propose we update to: "You will not receive notifications for this channel until channel mute is turned off."

File: i18n/en.json [hunk 2]
```diff
@@ -932,6 +932,34 @@
...
+  },
+  {
+    "id": "api.command_mute.success_unmute_direct_msg",
+    "translation": "Direct message channel is no longer muted."
```
Comment (esethna) [inline i18n/en.json]: Propose we update to: "This channel is no longer muted."

File: i18n/en.json [hunk 3]
```diff
@@ -932,6 +932,34 @@
...
+    "id": "api.command_mute.desc",
+    "translation": "Mutes the notifications for this channel"
+  },
+  {
+    "id": "api.command_mute.name",
+    "translation": "mute"
+  },
+  {
+    "id": "api.command_mute.error",
+    "translation": "An error occurred on muting {{.Channel}} occurred."
```
Comment (esethna) [inline i18n/en.json]: Propose we update to: "Could not find the channel {{.Channel}}."

File: i18n/en.json [hunk 4]
```diff
@@ -932,6 +932,34 @@
     "translation": "shrug"
   },
   {
+    "id": "api.command_mute.desc",
+    "translation": "Mutes the notifications for this channel"
```
Comment (esethna) [inline i18n/en.json]: Propose update to: "Turns off desktop, email and push notifications for the current channel or the [channel] specified."

Comment (esethna) [reply i18n/en.json]: Also the slash command title should be:


`/mute [channel]`

image

**Comment by chclaus** [comment]
Hej hej,

I've migrated the new mute property to mark_unread to follow the [advice](https://github.com/mattermost/mattermost-webapp/pull/196#issuecomment-339469343) of @hmhealey to preserve the current channel notification configuration.

File: i18n/en.json [hunk 5]
```diff
@@ -982,6 +982,34 @@
     "id": "api.command_shrug.name",
     "translation": "shrug"
   },
+  {
+    "id": "api.command_mute.desc",
+    "translation": "Mutes the notifications for this channel"
```
Comment (esethna) [inline i18n/en.json]: Please update to: `Turns off desktop, email and push notifications for the current channel or the [channel] specified.`

Comment (esethna) [reply i18n/en.json]: And please update the slash command title to `/mute [channel]`, 

image

Similar to the format of other slash commands, for example:
image

**Comment by esethna** [comment]
Thanks for the updates @chclaus, when you have a chance can you please help to address the above text changes? ^

**Comment by esethna** [comment]
@chclaus I'm going to move this PR out of the 4.5 release if that's okay? We're approaching the feature complete date of 11/29, after that this PR can be merged to master once the release candidate is cut early next month. Looking forward to it!

**Comment by chclaus** [comment]
Hej @esethna,

I've applied your I18n change requests, but stucking on the webapp part. Please have a [look](https://github.com/mattermost/mattermost-webapp/pull/196#issuecomment-347520906).

**Comment by crspeller** [comment]
@chclaus @esethna Going to close this because of inactivity. Please re-open when the PR is rebased and ready to go in.

**Comment by jasonblais** [comment]
Re-opening as this is being worked on in conjunction with the webapp PR. That PR has activity from earlier today https://github.com/mattermost/mattermost-webapp/pull/196

**Comment by esethna** [comment]
**On hold** until work is completed on the webapp PR.

**Comment by enahum** [comment]
@esethna this can be tested now...

I've made the push notifications to follow the `mute` channel preference and also changed the ephemeral messages for group messages to be equal to the direct messages

**Comment by jasonblais** [comment]
@enahum  looks like the build is failing

**Comment by enahum** [comment]
@jasonblais @AndersonWebStudio is taking a look cause it does not seem related to this PR

**Comment by enahum** [comment]
@jasonblais should be ok now

**Comment by enahum** [comment]
@chclaus please confirm the following:

There is no `mute` notification property but instead this is using the `mark_unread=mention` as the channel being muted, is that right?

**Comment by esethna** [comment]
Looks great @chclaus @enahum. Let's merge this so we can test webapp!

**Comment by enahum** [comment]
@esethna dev review is still pending

**Comment by esethna** [comment]
Yes after dev review of course. @jwilander @hmhealey can you please review when you get a chance?

**Comment by chclaus** [comment]
Hej guys,

thanks a lot. Especially to @enahum. It looks like you had more to do than I've expected. 

Best,
Christian

File: app/command_mute.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+		DisplayName:      T("api.command_mute.name"),
+	}
+}
+
+func (me *MuteProvider) DoCommand(a *App, args *model.CommandArgs, message string) *model.CommandResponse {
+	var channel *model.Channel
+	var noChannelErr *model.AppError
+
+	if channel, noChannelErr = a.GetChannel(args.ChannelId); noChannelErr != nil {
+		return &model.CommandResponse{Text: args.T("api.command_mute.error", map[string]interface{}{"Channel": channel.DisplayName}), ResponseType: model.COMMAND_RESPONSE_TYPE_EPHEMERAL}
```
Comment (cpanato) [inline app/command_mute.go]: This is correct? If there is an error, the channel will be nil, or am I missing something?

Comment (chclaus) [reply app/command_mute.go]: Hej @cpanato,

I think you're right. I'll fix this this afternoon (CET). Very thoughtful :)

Comment (chclaus) [reply app/command_mute.go]: Hej, see https://github.com/mattermost/mattermost-server/pull/8575

**Pull request merged**